src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilter.cs

```diff
@@ -24,12 +36,65 @@ public void OnTempDataSaving(ITempDataDictionary tempData)
                     var originalValue = kvp.Value;
 
                     var newValue = property.GetValue(Subject);
-                    if (newValue != null && newValue != originalValue)
+                    if (newValue != null && !newValue.Equals(originalValue))
                     {
                         tempData[Prefix + property.Name] = newValue;
                     }
                 }
             }
         }
+
+        public void OnActionExecuting(ActionExecutingContext context)
```

Comment:
Question for @rynowak : Should we consider async api on ITempDataDictionary & ITempDataProvider at this point( (yeah there would be breaking changes)? Currently these apis are sync (for obvious reason that TempData access is sync in view of user experience), but if we provide this binding support to properties directly, then we could load the data in an async way. (For example, if SessionStateTempDataProvider is used, it could load a session in async way using ISession's LoadAsync). Also we would attempt to load temp data for all actions on a controller even though some of them might use it, which increases chances of thread blockage.